### ISSUE TITLE
More granular verbosity

### ISSUE THREAD

**Issue by fstamour** [google/zx#306]
## Current Behaviour

When setting `$.verbose = false;` to hide the command that's going to be run, it also hides the command's output.

## Expected Behaviour

I would like a way to not print the command that's going to be run, but still show the command's output.

## Steps to reproduce

```$.shell = bash;```, because the example uses `select` which is bash-only.

```js
const listOfChoice = ['a', 'b', 'c'];
await $`select item in ${listOfChoice}
do
  echo $item
  break;
done`;
```

### `$.verbose = true;`

```txt
$ select item in a b c
> do
>   echo $item
>   break;
> done
1) a
2) b
3) c
#?
```

### `$.verbose = false;`

Will output nothing, but the shell will be waiting for a user input.

## Suggestions

> Just brainstorming here

1. add a `$.verboseCommand`, or 
2.  add a `$.hideOutput` configuration, or
3. change the type of  `$.verbose` (this would be backward-compatible, and extensible):

```typescript
type VerboseComponent = 'commands' | 'stdout' | 'stderr';
type Verbose = boolean | VerboseComponent | VerboseComponent[];

// example usage:
const verboseShowAll: Verbose = true;
const verboseShowAll2: Verbose = [ 'commands', 'stdout', 'stderr' ];
const verboseShowCommandsButNoOutput: Verbose = 'commands';
const verboseShowStderrOnly: Verbose = 'stderr';
const verboseShowOutputButNotCommand: Verbose = ['stderr', 'stdout' ];
const verboseShowNothing: Verbose = false;
```

## Specifications

- Version: 5.1.0
- Platform: any

**Comment by antonmedv** [comment]
No there is another feature for verbosity: https://github.com/google/zx#quiet

**Comment by fstamour** [comment]
> No there is another feature for verbosity: https://github.com/google/zx#quiet

"No", did you mean "Note"?

**Comment by antonmedv** [comment]
> "No", did you mean "Note"?

Yes)

**Comment by leoferreiralima** [comment]
i think can be a good ideia set a verboseLevel, like logs in general, info, warn, error, debug etc...

**Comment by jakebailey** [comment]
"hideOutput" is what I'm looking for, to emulate the `-x` flag in bash or similar.

Right now I have a terrible, terrible hack I can apply to trick `_run` into thinking the `ProcessPromise` is quiet after printing the command, which works, but makes me feel gross.

**Comment by antonmedv** [comment]
I think analog of `> /dev/null` is needed for zx.

Something like
```js
await $`verbose command`.pipe(null)
```

**Comment by jakebailey** [comment]
Maybe, but all I'm looking for is better control of printing; I don't want to lose the output, just not show it.

**Comment by fstamour** [comment]
btw, my original problem was that I was running a command that contained sensitive information that I really didn't want to show, but I also really needed to see the command's output.

I'm with @jakebailey I want/need better control of printing.

**Comment by antonmedv** [comment]
What about adding two methods like those?
- hideCommand
- hideOutput

**Comment by antonmedv** [comment]
For global config, I’m thinking about:

    $.verbose = 'only-commands'

Something like this.

**Comment by jakebailey** [comment]
I guess my mental model is that command printing and output printing are separate things entirely. Shells have the `-x` flag that prints out the commands executed, but outputs are controlled separately. In my case, I want to show my commands, but not necessarily all of their outputs (because I'm intending to capture `stdout` and process it.

That would boil down to having two "global" settings which are the defaults, one for the command logging, one for the output, then on each command, the ability to override those defaults with true/false.

For example, I might do something like this in bash:

```
#!/bin/bash

set -x # Print out all of the commands

someVar=$(some-command | jq | ...) # output is captured

git commit -m "$someVar" # shows output
```

Where in `zx` maybe I'm writing:

```js
import { $ } from `zx`;

$.showCommands = true;

const { stdout } = await $`some-command`.quiet() // ?

const something = stdout.split(/\r?\n/).map(....)...;

await $`git commit -m ${something}`
```

Or:

```js
import { $ } from `zx`;

$.showCommands = true;
$.showOutput = false;

const { stdout } = await $`some-command` // ?

const something = stdout.split(/\r?\n/).map(....)...;

await $`git commit -m ${something}`.showOutput()
```

But, I'm spitballing. Mainly, I just wanted to compare how a shell might do it; `zx` is different in that it always captures the various outputs but then prints them, unless you say `quiet`, in which case now you don't see the commands any more either.

**Comment by fstamour** [comment]
Don't forget there's also the use case where you want to hide stdout but not stderr.

**Comment by antonmedv** [comment]
Hide stdout via > /dev/null.

**Comment by jakebailey** [comment]
That works if you don't want to capture it, unless you mean some new piping mechanism.

**Comment by fstamour** [comment]
I doubt `> /dev/null` would work on windows, unless `zx` does some magic to translate to the right thing:tm:

**Comment by jakebailey** [comment]
I noticed that the code has had a massive update in the past week or so; it looks like I can get everything I need by doing:

```ts
$.verbose = 2; // default; prints all output
$.verbose = 1; // prints commands only
```

Although the readme hasn't yet been updated to say that this value can be numeric (the types are now generated from TS, so are always accurate). Unfortunately I'm not sure if I can upgrade, since some code changes seemed to have dropped Node 14 support (I'll file an issue about that, I guess).

Unrelated, but I was really confused as to why the code in the repo didn't match what was released, until I found the other v6 branch; you may want to switch GitHub's default branch to v6 (or merge v6 back into main and continue there). I almost filed an issue to say that the release code didn't match what was in the repo until I found it.

**Comment by antonmedv** [comment]
Yes, we are pushing right now to release v7 and update readme on front page.

**Comment by fstamour** [comment]
That doesn't fix my issue where I wanted to print only the command's output and not print the command...

**Comment by antonmedv** [comment]
For v7:
```js
$.log = entry => console.log(entry)

**Comment by fstamour** [comment]
that sounds good!

**Comment by RoyRao2333** [comment]
@antonmedv 

If I run

```ts
await $`tsx /xxx/yyy/z.ts my-flags`
```

and get output:

```zsh
$ tsx /xxx/yyy/z.ts my-flags

Hi! This is the final result!
```

How can I only allow my desired output `Hi! This is the final result!` and ignore the execution command (first line) `$ tsx /xxx/yyy/z.ts my-flags`?

Any help would be appreciated! :)

**Comment by antonmedv** [comment]
```js
```

**Comment by RoyRao2333** [comment]
@antonmedv 

Thanks for the reply, but if I set `$.verbose = false`, then I won't even see the `Hi! This is the final result!` output.

Anything else I can do?

**Comment by antonmedv** [comment]
Try this:

```js
$.verbose = false
await $`tsx /xxx/yyy/z.ts my-flags`.pipe(process.stdout)
```

Or this:

```js
import { LogEntry, log } from 'zx/core'

$.log = entry => {
  switch (entry.kind) {
    case 'cmd': break // skip
    default: log(entry)
  }
}
```

**Comment by RoyRao2333** [comment]
@antonmedv 

I tried the second approach and it's now working! You truly saved my day!

I got some questions though:

1. If I set `$.verbose`, `$.log` or any of `$`'s properties, does that mean that these properties will take effect whenever I execute `await $''` elsewhere? Or is there a specific scope?
2. According the first question, should I set `$`'s properties in a global scope or just before I execute `await $''`?

Thanks!
